jobspec2/hclutil/blockattrs.go

```diff
@@ -0,0 +1,175 @@
+package hclutil
+
+import (
+	"github.com/hashicorp/hcl/v2"
+	hcls "github.com/hashicorp/hcl/v2/hclsyntax"
+)
+
+// BlocksAsAttrs rewrites the hcl.Body so that hcl blocks are treated as
+// attributes when schema is unknown.
+//
+// This conversion is necessary for parsing task driver configs, as they can be
+// arbitrary nested without pre-defined schema.
+//
+// More concretely, it changes the following:
+//
+// ```
+// config {
+//   meta { ... }
+// }
+// ```
+// to
+//
+// ```
+// config {
+//   meta { ... }
+// }
+// ```
+func BlocksAsAttrs(body hcl.Body) hcl.Body {
+	if hclb, ok := body.(*hcls.Body); ok {
+		return &blockAttrs{body: hclb}
+	}
+	return body
+}
+
+type blockAttrs struct {
+	body hcl.Body
+
+	hiddenAttrs  map[string]struct{}
+	hiddenBlocks map[string]struct{}
+}
+
+func (b *blockAttrs) Content(schema *hcl.BodySchema) (*hcl.BodyContent, hcl.Diagnostics) {
+	bc, diags := b.body.Content(schema)
+	bc.Blocks = expandBlocks(bc.Blocks)
+	return bc, diags
+}
+func (b *blockAttrs) PartialContent(schema *hcl.BodySchema) (*hcl.BodyContent, hcl.Body, hcl.Diagnostics) {
+	bc, remainBody, diags := b.body.PartialContent(schema)
+	bc.Blocks = expandBlocks(bc.Blocks)
+
+	remain := &blockAttrs{
+		body:         remainBody,
+		hiddenAttrs:  map[string]struct{}{},
+		hiddenBlocks: map[string]struct{}{},
+	}
+	for name := range b.hiddenAttrs {
+		remain.hiddenAttrs[name] = struct{}{}
+	}
+	for typeName := range b.hiddenBlocks {
+		remain.hiddenBlocks[typeName] = struct{}{}
+	}
+	for _, attrS := range schema.Attributes {
+		remain.hiddenAttrs[attrS.Name] = struct{}{}
+	}
+	for _, blockS := range schema.Blocks {
+		remain.hiddenBlocks[blockS.Type] = struct{}{}
+	}
+
+	return bc, remain, diags
+}
+
+func (b *blockAttrs) JustAttributes() (hcl.Attributes, hcl.Diagnostics) {
+	body, ok := b.body.(*hcls.Body)
+	if !ok {
+		return b.body.JustAttributes()
+	}
+
+	attrs := make(hcl.Attributes)
+	var diags hcl.Diagnostics
+
+	if body.Attributes == nil && len(body.Blocks) == 0 {
```

Comment:
Does returning early get us anything here? Both will `x == nil` and `len(x) == 0` if empty, in which case there's nothing to range over.

Reply:
removing it - meant it to be fast-path optimization but seeing it now, it's not particularly useful.